python/ray/rllib/env/vector_env.py

```diff
@@ -101,3 +113,81 @@ def vector_step(self, actions):
     @override(VectorEnv)
     def get_unwrapped(self):
         return self.envs
+
+
+class _RemoteVectorizedGymEnv(_VectorizedGymEnv):
+    """Internal wrapper for gym envs to implement VectorEnv as remote workers.
+    """
+
+    def __init__(self, make_env, num_envs,
+                 action_space=None, observation_space=None):
+        _VectorizedGymEnv.__init__(self, make_env, [], num_envs,
+                                   action_space, observation_space)
+
+        for env in range(self.envs):
+            assert isinstance(env, ray.actor.ActorClass), \
+                "Your environment needs to be ray remote environment"
+
+    @override(_VectorizedGymEnv)
+    def vector_reset(self):
+        return ray.get([env.reset.remote() for env in self.envs])
+
+    @override(_VectorizedGymEnv)
+    def reset_at(self, index):
+        return ray.get(self.envs[index].reset.remote())
+
+    @override(_VectorizedGymEnv)
+    def vector_step(self, actions):
+        step_outs = ray.get([env.step.remote(act)
+                             for env, act in zip(self.envs, actions)])
+
+        obs_batch, rew_batch, done_batch, info_batch = [], [], [], []
+        for obs, rew, done, info in step_outs:
+            obs_batch.append(obs)
+            rew_batch.append(rew)
+            done_batch.append(done)
+            info_batch.append(info)
+        return obs_batch, rew_batch, done_batch, info_batch
```

Comment:
I think we need some way of destroying these remote actors when the agent is stopped, otherwise the actors may be leaked. (I haven't checked if they actually are).

Reply:
Problem is, I don't know how we can do that if the framework doesn't start ray workers. And then we come back to the issues we were talking about earlier...

But I'm not sure, seems that those are getting closed (they do not exist after keyboard interrupt). But other ugly kinds of stuff are happening - everything hangs if you request more environments than you can schedule on ray, etc (because it is out of the framework rllib does not handle it).

Reply:
Makes sense for now, I think you only run into the issue if running multiple experiments on the same cluster.

Re: resources: I think that can be resolved with the right resource requests (at least when running in Tune):

Basically, if you have remote envs enabled, then the agent should request `max(1, num_workers) * num_envs_per_worker * num_remote_envs` extra_cpu in default_resource_request(). That way you won't run into this resource allocation deadlock. However this only applies if running in Tune.

Alternatively, I think it would also work to create these remote envs with `num_cpus=0`.

Reply:
Yeah, I figured out those ways around the issue, just mentioned that we don't control the remote environments and we don't even raise an exception if there are no resources when building them, it just hangs.